### ISSUE TITLE
False positive for `package_names` lint for name without underscores.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/linter#2803]
**Describe the issue**

The package `async` gets a warning about its package name not being lowercase-with-underscores.

**To Reproduce**

Make a package with a single-identifier name. Enable recommended lints.

**Expected behavior**

No warning. Single-identifier package names are common (`characters`, `test`, `flutter`, etc.).

**Additional context**

The RegExp in `utils.dart` is 
```dart
final _lowerCaseUnderScoreWithLeadingUnderscores =
    RegExp(r'^(_)*([a-z]+([_]?[a-z0-9]+)*)+$');
```

The final `+` should possibly be `*` instead.

**Comment by lrhn** [comment]
That RegExp worries me. The nested `+)*)+` looks like something which can cause catastrophic backtracking on almost compatible strings.

Checking ... yep, it's bad. The time for the following code seems to grow exponentially with the length of the string (~O(3<sup>|string|</sup>)), until it reaches i == 20, then it just doesn't complete: 
```dart
var re = RegExp(r'^(_)*([a-z]+([_]?[a-z0-9]+)*)+$');
void main() async {
  for (var i = 0; i < 200; i++) {
    var s = "a" * i + "Z";
    print(s);
    var sw = Stopwatch()..start();  
    print("$i: ${re.hasMatch(s)} : ${sw.elapsedMilliseconds}");
    await Future.delayed(Duration(milliseconds: 10));
  }
}
```
so we have a potential denial of service just from writing a 20-char name.

Let's rewrite that.
The specification would be:
* a sequence of lower case letters, digits and underscores,
* starting with zero or more underscores and then a lower case letter,
* not containing two adjacent underscores later in the string,
* and not ending with an underscore.
The captures appear to be unnecessary, so:
```dart
   RegExp(r'^_*[a-z](?:_?[a-z\d])*$');
```

I'll make a CL to fix some of the other dangerous RegExps.

**Comment by lrhn** [comment]
I notice that the original problem should already be fixed according to https://github.com/dart-lang/linter/issues/2790, so I'll just concentrate on the RegExps.

**Comment by pq** [comment]
Thanks!
